File: articles/secure-computation-and-privacy-preserving-algorithms/cpp/privacy_preserving_demo.cpp
#include <iostream>
#include <vector>

struct ClientUpdate {
    int examples;
    double weight;
};

double federated_average(const std::vector<ClientUpdate>& updates) {
    int total = 0;
    double weighted = 0.0;
    for (const auto& update : updates) {
        total += update.examples;
        weighted += update.examples * update.weight;
    }
    return total == 0 ? 0.0 : weighted / total;
}

int main() {
    std::vector<ClientUpdate> updates = {{100, 0.42}, {240, 0.55}, {160, 0.49}};
    std::cout << "federated average weight=" << federated_average(updates) << "\n";
    return 0;
}

File: articles/secure-computation-and-privacy-preserving-algorithms/rust/privacy_preserving_demo.rs
struct ClientUpdate {
    examples: f64,
    weight: f64,
}

fn federated_average(updates: &[ClientUpdate]) -> f64 {
    let total: f64 = updates.iter().map(|u| u.examples).sum();
    let weighted: f64 = updates.iter().map(|u| u.examples * u.weight).sum();
    if total == 0.0 { 0.0 } else { weighted / total }
}

fn main() {
    let updates = [
        ClientUpdate { examples: 100.0, weight: 0.42 },
        ClientUpdate { examples: 240.0, weight: 0.55 },
        ClientUpdate { examples: 160.0, weight: 0.49 },
    ];
    println!("federated average weight={:.6}", federated_average(&updates));
}
